src/zope/interface/interface.py

```diff
@@ -875,9 +875,14 @@ def direct(self, name):
     def queryDescriptionFor(self, name, default=None):
         return self.get(name, default)
 
-    def validateInvariants(self, obj, errors=None):
+    def validateInvariants(self, obj, errors=None, seen=None):
```

Comment:
The implementation seems fine to me. Invariants might not be hashable, but that seems unlikely.

`validateInvariants` is a public method with a documented signature, though, and I'm not sure I like adding a public argument to that. The documentation and introspection by things like IDEs will disagree. I've seen this handled in two different ways. One is to name the argument with a leading underscore, indicating its for internal usage.
```suggestion
    def validateInvariants(self, obj, errors=None, _seen=None):
```

The other is to move the real work to an internal function:
```suggestion
    def validateInvariants(self, obj, errors=None):
        return self.__validateInvariants(obj, errors, set())
```
This has the advantage of not needing to check that `seen` is None at the start of the function. (We'd use `base.__validateInvariants(obj, errors, seen)` below too, of course.)

It's possible, (but not likely), that someone has subclassed `InterfaceClass` and overridden `validateInvariants` or used `@interfacemethod` to write their own `validateInvariants` method. The current implementation, as well as the first alternate form, could break such classes. I think the second alternate form (internal implementation) would be safe, but could wind up bypassing the custom methods. 

Thoughts?

Reply:
Thank you for your thorough explanation @jamadden. My first instinct would be to follow the `_seen` implementation, but moving the code to  a `__validateInvariants` [looks cleaner](https://github.com/zopefoundation/zope.interface/pull/215/commits/64fb805b3b2a6536c4874b4f0c3fbbbf4bea9e23), so I am okay with the second approach.